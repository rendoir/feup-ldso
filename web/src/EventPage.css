div#page-event {
    margin:  1em 4em 2em 4em;
    padding: 1em 1em;
    background: white;
}

div.row.event-page-header{
    margin: 2em 0;
}

div.event-title {
    padding-left: 2em;
}

div.event-title h2{
    text-align: left;
}

div.event-page-buttons {
    padding: 0 3em 0 0;
    text-align: right;
}

div.row.event-page-body {
    margin: 2em 0;
    padding: 0 2em;
}

div.row.event-page-body > div:first-of-type {
    text-align: left;
}

img.event-image {
    margin: 1em auto 1em 0;
    max-width: 100%;
    max-height: 300px;
}

p.event-description {
    margin: 1em 0;
    white-space: pre-wrap;
    text-align: justify;
}

span.category-span {
    background-color: #4a4a4a;
    border-radius: 10px;
    color: white;
    padding: 0.5em;
    margin: 0 0.5em;
}

div.row.event-page-info {
    margin: 2em 0;
    text-align: left;
}

div.row.event-page-info > div:first-of-type{
    padding-left: 4em;
}

div.row.event-page-info div div {
    margin: 1em 0;
}


div.div-border {
    border-right: 1px solid black;
}



h4.display-inline {
    display: inline;
}

button.btn {
    background-color: inherit;
    border-color: inherit;
    color: #4a4a4a;
}

.btn.btn-link.btn-primary {
    color: #4a4a4a;
    border-color: inherit;
}

.btn.btn-primary:hover {
    background-color: #4a4a4a;
    color: white;
}

div.event-page-button-map a {
    transform-origin: center center;
    position: absolute;
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    height: 35px;
    width: 120px;
    margin: auto auto;
}

div.loading {
    transform-origin: center center;
    position: absolute;
    margin-left: auto;
    margin-right: auto;
    left: 0;
    right: 0;
    top: 45%;
    width: 120px;
    height: 120px;
}

.loader {
    border: 16px solid #f3f3f3; /* Light grey */
    border-top: 16px solid #32a794; /* Blue */
    border-radius: 50%;
    transform-origin: center center;
    height: 100%;
    width: 100%;
    animation: spin 2s linear infinite;
}

@keyframes spin {
    0% { transform: rotate(0deg); }
    100% { transform: rotate(360deg); }
}