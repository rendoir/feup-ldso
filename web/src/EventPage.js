import React, { Component } from 'react';
import { Col, Row, Alert, Image, Button, Breadcrumb } from 'react-bootstrap';
import { Redirect, Link } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import axios from 'axios';
import swal from 'sweetalert';
import './EventPage.css';

function getTokenFromCookie() {
    let token = document.cookie.split("access_token=")[1];
    return token;
}

class EventPage extends Component {

    constructor(props) {
        super(props);

        this.state = {
            id: props.match.params.id,
            title: "",
            title_english: "",
            description: "",
            description_english: "",
            entity_id: null,
            entity: null,
            end_date_nonformat: "",
            start_date_nonformat: "",
            end_date: "",
            start_date: "",
            location: "",
            price: null,
            categories: [],
            image: null,
            alertType: null,
            alertMessage: null,
            redirect: false,
            errorLoadingImage: true
        };

        this.deleteEvent = this.deleteEvent.bind(this);
        this.deleteEventConfirmed = this.deleteEventConfirmed.bind(this);
        this.getDefaultImage = this.getDefaultImage.bind(this);
    }

    componentDidMount() {
        axios.get('http://' + process.env.REACT_APP_API_URL + ':3030/events/' + this.state.id)
            .then((res) => {

                if (res.data === "" || Object.keys(res.data).length === 0) {
                    this.setState({ alertType: "danger", alertMessage: "Este evento não existe." });
                } else {

                    let date = new Date(res.data.start_date);
                    let dd = date.getDate();
                    let mm = date.getMonth() + 1;
                    let yyyy = date.getFullYear();
                    if (dd < 10) {
                        dd = '0' + dd;
                    }
                    if (mm < 10) {
                        mm = '0' + mm;
                    }
                    let stringDateStart = dd + "/" + mm + "/" + yyyy + " " + date.getHours() + ":" + (date.getMinutes() < 10 ? '0' : '') + date.getMinutes();

                    let stringDateEnd = null;
                    if (res.data.end_date !== null) {
                        date = new Date(res.data.end_date);
                        dd = date.getDate();
                        mm = date.getMonth() + 1;
                        yyyy = date.getFullYear();
                        if (dd < 10) {
                            dd = '0' + dd;
                        }
                        if (mm < 10) {
                            mm = '0' + mm;
                        }
                        stringDateEnd = dd + "/" + mm + "/" + yyyy + " " + date.getHours() + ":" + (date.getMinutes() < 10 ? '0' : '') + date.getMinutes();
                    }

                    this.setState({
                        title: res.data.title,
                        title_english: res.data.title_english,
                        description: res.data.description,
                        description_english: res.data.description_english,
                        start_date_nonformat: res.data.start_date,
                        end_date_nonformat: res.data.end_date,
                        start_date: stringDateStart,
                        end_date: stringDateEnd,
                        location: res.data.location,
                        price: res.data.price,
                        entity: res.data.entity,
                        entity_id: res.data.entity_id,
                        categories: res.data.categories,
                        image: "http://" + process.env.REACT_APP_API_URL + ":3030/" + this.state.id + "?" + Date.now()
                    });

                }

            })
            .catch(() => {
                this.setState({ alertType: "danger", alertMessage: "Ocorreu um erro. Por favor tente novamente." });
            });
    }

    deleteEvent() {
        swal({
            title: "Tem a certeza que quer apagar o evento?",
            icon: "warning",
            buttons: true,
            dangerMode: true
        })
            .then((isConfirm) => {
                if (isConfirm)
                    this.deleteEventConfirmed();
            });
    }

    deleteEventConfirmed() {
        let self = this;
        axios.delete('http://' + process.env.REACT_APP_API_URL + ':3030/', {
            data: {
                id: self.state.id
            },
            headers: { 'Authorization': "Bearer " + getTokenFromCookie() }
        })
            .then(() => this.setState({ redirect: true }))
            .catch(() => self.setState({ alertType: "danger", alertMessage: "Um erro ocorreu a apagar o evento. Tente mais tarde." }));
    }

    getDefaultImage() {
        if (this.state.errorLoadingImage) {
            let image = document.querySelector('img.event-image');
            image.src = "/default.png";
            this.setState({ errorLoadingImage: false });
        }
    }


    render() {

        if (this.state.redirect) {
            return <Redirect to="/events" push />;
        }

        let alertElement;
        if (this.state.alertMessage !== null) {
            alertElement = (
                <Row>
                    <Col sm={4} md={2}>

                    </Col>
                    <Col sm={5} md={8}>
                        <Alert className={this.state.alertType}>
                            {this.state.alertMessage}
                        </Alert>
                    </Col>
                    <Col sm={4} md={2}>

                    </Col>Ocorreu um erro. Por favor tente novamente.
                </Row>);
        }

        let categories = this.state.categories.map((cat, i) => {
            return <span key={i} className="category-span">{cat.name}</span>;
        });


        if (this.state.title === "") {
            return (
                <div className="loading">
                    <div className="loader"></div>
                </div>
            );
        } else {

            return (
                <div>
                    {alertElement}
                    <Row>
                        <Col sm={4} md={2}>

                        </Col>
                        <Col sm={5} md={8}>
                            <Breadcrumb>
                                <Breadcrumb.Item>
                                    <Link to={`/events`}>
                                        Eventos
                                    </Link>
                                </Breadcrumb.Item>
                                <Breadcrumb.Item active>{this.state.title}</Breadcrumb.Item>
                            </Breadcrumb>
                        </Col>
                        <Col sm={4} md={2}>

                        </Col>
                    </Row>
                    <div id="page-event">
                        <Row className="event-page-header">
                            <Col sm={2}>
                            </Col>
                            <Col sm={8} className="event-title">
                                <h2>{this.state.title} / {this.state.title_english}</h2>
                                <div className="event-page-entities">
                                    <span>Entidade: </span><span className="event-page-entities-name">{this.state.entity.initials}</span>
                                </div>
                            </Col>
                            <Col sm={2}>
                            </Col>
                        </Row>
                        <Row className="event-page-body">
                            <Col sm={4}>
                                <Image
                                    src={this.state.image}
                                    className="event-image"
                                    onError={this.getDefaultImage}
                                />

                            </Col>
                            <Col sm={8}>
                                <p className="event-description">{this.state.description}</p>
                                <p className="event-description">{this.state.description_english}</p>
                            </Col>
                        </Row>
                        <Row className="event-page-info">
                            <Col sm={5}>
                                <div className="event-date-hour">
                                    <h4 className="display-inline">Dia/Hora: </h4>
                                    <span>{this.state.start_date}</span>
                                    <span> {this.state.end_date !== null ? "-" + this.state.end_date : ""}</span>
                                </div>
                                <div className="event-location">
                                    <h4 className="display-inline">Localização: </h4>
                                    <span>{this.state.location}</span>
                                    <Button
                                        className="location-button"
                                        href={`https://www.google.com/maps/search/?api=1&query=${encodeURI(this.state.location)}`}
                                        target="_blank"
                                        rel="noopener noreferrer">
                                        <FontAwesomeIcon icon="map-marked-alt" />
                                    </Button>
                                </div>
                            </Col>
                            <Col sm={5} className="event-page-ent-cat">
                                <div className="event-page-categories">
                                    <h4 className="display-inline">Categorias: </h4>
                                    {categories}
                                </div>
                                <div className="event-page-price">
                                    <h4 className="display-inline">Preço: </h4>
                                    <span>{this.state.price}€</span>
                                </div>
                            </Col>
                            <Col sm={2} className="event-page-buttons">
                                <Link className="btn btn-link btn-primary" to={{
                                    pathname: `/events/${this.state.id}/edit`,
                                    state: { event: this.state }
                                }}>
                                    <FontAwesomeIcon icon="edit" />
                                </Link>
                                <Button className="delete-button" onClick={this.deleteEvent}>
                                    <FontAwesomeIcon icon="trash-alt" />
                                </Button>
                            </Col>
                        </Row>
                    </div>
                </div>
            );

        }

    }
}

export default EventPage;
