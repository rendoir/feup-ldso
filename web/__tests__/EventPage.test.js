import React from 'react';
import rendered from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import EventPage from '../src/EventPage';
import { mount } from 'enzyme';
import sinon from 'sinon';
import axios from 'axios';
import swal from 'sweetalert';
import MockAdapter from 'axios-mock-adapter';

var mockAxios = new MockAdapter(axios);
var props = {
    match: {
        params: {
            id: 1
        }
    }
};

describe("Render", () => {

    it('renders event page', () => {

        mockAxios.onGet().reply(200, {
            id: 1,
            title: 'Title',
            description: 'description',
            start_date: '2018-10-27 11:11:00',
            end_date: '2018-10-28 11:11:00',
            entity_id: 1,
            location: 'Porto',
            price: 10,
            entity: {
                id: 1,
                name: 'Faculdade de Engenharia da Universidade do Porto',
                initials: 'FEUP',
                description: 'description',
                color: '#ffffff',
                color2: '#ffffff',
                location: 'Porto'
            },
            categories: [
                {
                    id: 1,
                    name: 'Category1',
                    description: 'Description'
                },
                {
                    id: 2,
                    name: 'Category2',
                    description: 'Description'
                }
            ]
        });


        const event = rendered.create(
            <BrowserRouter>
                <EventPage {...props} />
            </BrowserRouter>);
        let treeList = event.toJSON();
        expect(treeList).toMatchSnapshot();
    });

});


describe("Check ComponentDidMount", () => {

    it("/GET page info - Success", async(done) => {

        mockAxios.onGet().reply(200, {
            id: 1,
            title: 'Title',
            description: 'description',
            start_date: '2018-10-27 11:11:00',
            end_date: '2018-10-28 11:11:00',
            entity_id: 1,
            location: 'Porto',
            price: 10,
            entity: {
                id: 1,
                name: 'Faculdade de Engenharia da Universidade do Porto',
                initials: 'FEUP',
                description: 'description',
                color: '#ffffff',
                color2: '#ffffff',
                location: 'Porto'
            },
            categories: [
                {
                    id: 1,
                    name: 'Category1',
                    description: 'Description'
                },
                {
                    id: 2,
                    name: 'Category2',
                    description: 'Description'
                }
            ]
        });

        let wrapper = await mount(
            <BrowserRouter>
                <EventPage {...props} />
            </BrowserRouter>
        );

        let wrapperPage = await wrapper.find(EventPage).first();

        await wrapperPage.update();


        setImmediate(() => {
            expect(wrapperPage.state().id).toEqual(1);
            expect(wrapperPage.state().title).toEqual('Title');
            expect(wrapperPage.state().description).toEqual('description');
            expect(wrapperPage.state().start_date).toEqual('27/10/2018 11:11');
            expect(wrapperPage.state().end_date).toEqual('28/10/2018 11:11');
            expect(wrapperPage.state().entity_id).toEqual(1);
            expect(wrapperPage.state().location).toEqual('Porto');
            expect(wrapperPage.state().price).toEqual(10);
            expect(wrapperPage.state().entity).toEqual({
                id: 1,
                name: 'Faculdade de Engenharia da Universidade do Porto',
                initials: 'FEUP',
                description: 'description',
                color: '#ffffff',
                color2: '#ffffff',
                location: 'Porto'
            });
            expect(wrapperPage.state().categories.length).toEqual(2);
            done();
        });

    });

    it("/GET page info - Doens't exist", async(done) => {

        mockAxios.onGet().reply(200, {});

        let wrapper = await mount(
            <BrowserRouter>
                <EventPage {...props} />
            </BrowserRouter>
        );

        let wrapperPage = await wrapper.find(EventPage).first();

        await wrapperPage.update();


        setImmediate(() => {
            expect(wrapperPage.state().alertType).toEqual("danger");
            expect(wrapperPage.state().alertMessage).toEqual("Este evento nÃ£o existe.");
            done();
        });

    });

    it("/GET page info - Error", async(done) => {

        mockAxios.onGet().reply(400);

        let wrapper = await mount(
            <BrowserRouter>
                <EventPage {...props} />
            </BrowserRouter>
        );

        let wrapperPage = await wrapper.find(EventPage).first();

        await wrapperPage.update();

        setImmediate(() => {
            expect(wrapperPage.state().alertType).toEqual('danger');
            expect(wrapperPage.state().alertMessage).toEqual('Ocorreu um erro. Por favor tente novamente.');
            done();
        });

    });

});

describe("Delete Event", () => {

    it("Check if swal appears when delete button is clicked", async() => {

        mockAxios.onGet().reply(200, {
            id: 1,
            title: 'Title',
            description: 'description',
            start_date: '2018-10-27 11:11:00',
            end_date: '2018-10-28 11:11:00',
            entity_id: 1,
            location: 'Porto',
            price: 10,
            entity: {
                id: 1,
                name: 'Faculdade de Engenharia da Universidade do Porto',
                initials: 'FEUP',
                description: 'description',
                color: '#ffffff',
                color2: '#ffffff',
                location: 'Porto'
            },
            categories: [
                {
                    id: 1,
                    name: 'Category1',
                    description: 'Description'
                },
                {
                    id: 2,
                    name: 'Category2',
                    description: 'Description'
                }
            ]
        });

        let spy = sinon.spy(swal);

        const wrapper = await mount(
            <BrowserRouter>
                <EventPage {...props} />
            </BrowserRouter>
        );


        let wrapperPage = await wrapper.find(EventPage).first();

        await wrapperPage.instance().forceUpdate();
        await wrapperPage.update();

        const button = await wrapperPage.update().find('button.delete-button').first();
        button.simulate('click');

        expect(spy.calledOnce);


    });

    it("Check if deletion was successfull", async(done) => {

        let state = {
            id: 1,
            title: 'Title',
            description: 'description',
            start_date: '2018-10-27 11:11:00',
            end_date: '2018-10-28 11:11:00',
            entity_id: 1,
            location: 'Porto',
            price: 10,
            entity: {
                id: 1,
                name: 'Faculdade de Engenharia da Universidade do Porto',
                initials: 'FEUP',
                description: 'description',
                color: '#ffffff',
                color2: '#ffffff',
                location: 'Porto'
            },
            categories: [
                {
                    id: 1,
                    name: 'Category1',
                    description: 'Description'
                },
                {
                    id: 2,
                    name: 'Category2',
                    description: 'Description'
                }
            ],
            alertType: null,
            alertMessage: null,
            redirect: false,
            errorLoadingImage: true
        };

        mockAxios.onDelete().reply(200);

        const wrapper = await mount(
            <BrowserRouter>
                <EventPage {...props} />
            </BrowserRouter>);

        let wrapperPage = await wrapper.find(EventPage).first();

        await wrapperPage.setState(state);
        await wrapperPage.update();

        await wrapperPage.instance().deleteEventConfirmed();

        setImmediate(() => {
            expect(wrapperPage.state().redirect).toEqual(true);
            done();
        });


    });

    it("Check if deletion wasn't successfull", async(done) => {

        mockAxios.onDelete().reply(400);

        const wrapper = await mount(
            <BrowserRouter>
                <EventPage {...props} />
            </BrowserRouter>);

        let wrapperPage = await wrapper.find(EventPage).first();

        await wrapperPage.instance().deleteEventConfirmed();

        await wrapperPage.update();

        setImmediate(() => {
            expect(wrapperPage.state().alertType).toEqual("danger");
            expect(wrapperPage.state().alertMessage).toEqual("Um erro ocorreu a apagar o evento. Tente mais tarde.");
            done();
        });
    });

});

describe("Get Image", () => {

    it("/GET Event Image - Error", async(done) => {

        mockAxios.onGet().reply(400);

        let wrapper = await mount(
            <BrowserRouter>
                <EventPage {...props} />
            </BrowserRouter>
        );

        let wrapperPage = await wrapper.find(EventPage).first();

        wrapperPage.instance().getDefaultImage = jest.fn();
        let spy = sinon.spy(wrapperPage.instance().getDefaultImage);

        await wrapperPage.update();

        setImmediate(() => {
            expect(spy.calledOnce);
            done();
        });

    });

    it("/GET Event Image - Check getDefaultImage", async(done) => {

        let wrapper = await mount(
            <BrowserRouter>
                <EventPage {...props} />
            </BrowserRouter>
        );

        let wrapperPage = await wrapper.find(EventPage).first();

        wrapperPage.setState({ errorLoadingImage: true });

        let element = document.createElement('img');
        element.src = '';
        document.querySelector = jest.fn().mockReturnValue(element);

        wrapperPage.instance().getDefaultImage();

        await wrapperPage.update();

        setImmediate(() => {
            expect(wrapperPage.state().errorLoadingImage).toEqual(false);
            done();
        });

    });

});
