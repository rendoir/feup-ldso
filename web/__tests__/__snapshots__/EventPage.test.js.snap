// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render renders event page 1`] = `
<div
  className="loading"
>
  <div
    className="loader"
  />
</div>
`;

exports[`renders elements 1`] = `
<nav
  className="navbar navbar-expand navbar-light"
  id="navbar-custom"
>
  <a
    aria-current={null}
    className="navbar-brand"
    href="/events"
    onClick={[Function]}
  >
    <img
      className=""
      id="logo"
      src="/logotipo_agenda.png"
    />
  </a>
  <div
    aria-expanded={null}
    className="navbar-collapse collapse"
    id="responsive-navbar-nav"
  >
    <div
      className="navbar-nav"
      onKeyDown={[Function]}
    />
    <div
      className="navbar-nav"
      onKeyDown={[Function]}
    />
  </div>
</nav>
`;

exports[`renders elements 2`] = `null`;
